### PR TITLE
MCP: API Guidelines

### PR DESCRIPTION
**Pull request by KodrAus**
This MCP is to try rework the API Guidelines discussion a bit to separate active discussion from actual guideline proposals, and to see if there are changes we’d like to make leading up to a 2021 edition.

[Rendered](https://github.com/rust-lang/libs-team/blob/mcp/api-guidelines/mcps/0006-api-guidelines.md)

### CODE REVIEW

**Comment by BurntSushi** [comment]
Thanks for writing this up! I have a couple of comments.

Firstly, I think one of the existing issues with the API guidelines is that it's not clear (at least to me) when a decision has been made and the guidelines should be updated. In rust-lang/rust, we use FCP to do that. Could we do the same in this repo?

Secondly, and perhaps related to the first, I am not sure on this phrasing:

> Once discussion on Zulip has settled on a guideline then a PR to `rust-lang/api-guidelines` should be opened to propose it.

Does this mean that discussion should continue on that PR? From this phrasing, I would worry a bit that folks might think that if folks settle on something in Zulip then the PR is a foregone conclusion. But I think there should be room for asynchronous feedback and folks should expect that when opening a PR. The wording here does say "propose," but I actually missed that on the first read through. Maybe we could clarify this a bit?

**Comment by KodrAus** [comment]
Thanks @BurntSushi! :bow:

I haven't looked at them really closely yet, but I wonder if GitHub Discussions might be a good fit here rather than Zulip.

I think the wording there is a little unclear, isn't it. I was imagining discussion would naturally enumerate all the considerations around a particular topic and a member of Libs would then suggest we'd be open to reviewing a proper proposal. That does leave the question open for where discussion should continue afterwards.

What if we introduced a bit more of a two-tier book? The first one could be like "API Considerations", which would be a bit of a grab-bag of considerations that have different perspectives we might have (or intend to) fully consider. PRs to those considerations expanding their content would be accepted. If a consideration reached a point that we could make a concrete guideline out of it then we'd create an issue requesting it, and FCP a PR that introduced it?

To use MSRVs as an example, the whole process might look something like:

- A consideration about testing a specific compiler version in CI so you notice bumps, keeping an eye on the versions of Rust available in package managers etc. Some strategies folks have used for treating version bumps without making any specific recommendations, maybe something on testing with minimal dependency version resolution instead of the default?
- At some point in the future, Cargo introduces a package field for a minimal compiler version (hypothetically). We now look at suggesting a guideline to include a minimal version field in your package before `1.0`.
- Either an issue is opened to promote the consideration or we just get around to PRing a new guideline sometime. In either place we do an FCP to introduce it which may generate some discussion to feed back to the consideration or get merged.

These considerations could live alongside guidelines so links to them don't change, but they're flagged as open discussion points.

What do you think?

**Comment by BurntSushi** [comment]
I would worry about that process being a bit heavyweight and also suffering from "considerations" piling up and languishing. But I'm not opposed to trying it. Mainly, I think the value of the API guidelines is that it gives a specific set of concrete suggestions that don't require folks to toil and make their own decisions. (And I don't mean that in a bad way, it's just sometimes the thing we want.) With that said, I don't necessarily think it's a bad idea. Maybe listing "considerations" as "a step below a recommendation" or something similar might be a nice way to attract more attention/discussion to it, especially if it linked to open issue on this tracker as a way of inviting more discussion on it.

I feel like that's mostly orthogonal to my questions though? I think the main thing I'm interested in is specifically how _decisions_ are made. If it's like the FCP process we have, then I'm happy with that, because it gives a chance for asynchronous discussion. But otherwise having discussions on zulip or what not before that seems fine to me, as long as they can be linked from the FCP. It sounds like FCP is what you have in mind though.

I'm not sure about letting considerations bypass FCP though. To be honest, it's hard to do know what the right balance is without actually trying something. I would probably start conservative (FCP more things) because I think it's easier to say "nah we don't need FCP for this" than it is to say, "we didn't need it before, but now let's require it before merging changes like foo."

**Comment by KodrAus** [comment]
Ah yes, I was thinking we'd FCP any new guidelines.

I was thinking there are really two resources in the API Guidelines. There's the guidelines themselves, and there's the discussion about patterns and practices. I've been trying to make the latter a bit more usable by moving discussions off issues, but we need a clear way of making decisions to link them back up again so discussions lead to actionable guidelines that we all think are really broadly applicable. I agree that FCPs seem like the right tool for that.

With considerations, I was thinking we'd have a place to capture different perspectives over time before trying to condense that into a guideline if there's a concrete recommendation we can make from it. But that's something we could just do in discussions. If we're able to give GitHub discussions a go we might be able to issue FCPs directly on discussions for guidelines we'd like to introduce.

**Comment by BurntSushi** [comment]
Sounds worth giving a shot. Would be good to inject some life into the API guidelines project.

**Comment by KodrAus** [comment]
We now have GitHub Discussions enabled on the API Guidelines: https://github.com/rust-lang/api-guidelines/discussions

Unfortunately it doesn't look like we'll be able to issue FCPs in them yet because there isn't a public API for discussions, but I imagine that will become possible sometime soon.

**Comment by KodrAus** [comment]
Alrighty, I’ve spent some time with the Discussions feature and have updated the text here to suggest we use them instead of Zulip, and makes clear how and where we expect decisions to be made.

What do you think @BurntSushi?

**Comment by BurntSushi** [comment]
@KodrAus LGTM! Thanks! :)

**Comment by KodrAus** [comment]
There's also [some discussion](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Rust.20Design.20Patterns) now about bringing the [Patterns](https://github.com/rust-unofficial/patterns) in to `rust-lang` as a complement to the Guidelines (so we can shift patterns-related discussion over to it). I don't think that would affect us FCPing guidelines changes, and I don't think we'd need to FCP patterns (at least not at this stage) but it seems worth dropping a comment here about it!

File: mcps/0006-api-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# API Guidelines
+
+## Writing a 2021 Edition of the API Guidelines
+
+Some patterns have changed and new idioms have emerged since we last spent time on the API Guidelines.
+There are some new language features like `impl Trait` and `async` / `await` that haven't been fully considered in the existing guidelines yet.
+
+As of this writing, it's been about two years since any substantial new content was added to the API Guidelines.
+That's not necessarily an issue, we could consider the API Guidelines "done" for the 2018 language edition.
+Coming up to 2021 we should revist existing recommendations and see if any new ones have come up that make sense for the new edition.
```
Comment (camelid) [inline mcps/0006-api-guidelines.md:10]: Typo:

```suggestion
Coming up to 2021 we should revisit existing recommendations and see if any new ones have come up that make sense for the new edition.
```

**Comment by KodrAus** [comment]
This is the current state of the Patterns and Guidelines charter that would subsume this MCP: https://hackmd.io/ENxEJyMhSzirz1U1XQHzbw

**Comment by KodrAus** [comment]
I no longer have the bandwidth to carry this forward but if somebody else would like to pick it up sometime in the future then please feel free!

**Pull request closed without merging**